### DISCUSSION TITLE
Process Manager - longer term state

### DISCUSSION THREAD

**Discussion by dazinator** [rebus-org/Rebus#1115] [General]
I've created a workflow engine where workflows are defined in yaml files and when the workflow is executed, I use rebus saga to orchestrate it through to completion.

Right now the when the workflow completes the saga completes and the state is removed.

In the future I want to add a UI which will show a list of completed workflows and the results for each step etc.

My plan to deliver this is, is for the saga to send a message when the workflow "completes" - where the message contains the final state of the workflow - in order to replicate the final state out of rebus saga and allow the saga to be deleted in rebus saga tables.

- any large data will be databus tickets, so will keep the size of this state on the workflow completed message down as much as possible.

Then I'll implement a handler for this workflow completed message that will materialise this state into tables for the purposed of my UI application to query. This will be completely seperate from the rebus saga tables where the information will be deleted.

In future I also want to introduce a UI to show currently running workflows. To achieve this I'll update the rebus saga so that it sends messages or notifications when workflows are started and other interesting things happen. I can then implement handlers for those events to keep my tables for the UI in "sync" with whats running.

I haven't thought through all the details just yet, but curious to hear if anyone has any input on such a design and whether this seems sensible approach

**Comment by mookid8000** [comment]
Another option you could maybe put to use is Rebus' built-in saga auditing – it will simply store a snapshot of each saga data in the configured snapshot storage (an implementation of [`ISagaSnapshotStorage`](https://github.com/rebus-org/Rebus/blob/master/Rebus/Auditing/Sagas/ISagaSnapshotStorage.cs))

You can enable it like this (in this case using Rebus.MongoDb):

```csharp
services.AddRebus(
	configure => configure
		.(...)
        .Options(o => {
               o.EnableSagaAuditing().StoreInMongoDb(db, "Snapshots");
        })		
);
```

**Reply by dazinator** [reply]
Interesting, I will look into that, thanks.

**Comment by DavidEggen** [comment]
@dazinator have you had any luck? I am wanting to do something similar.

**Reply by dazinator** [reply]
Nope sorry not yet. It will be a while before I visit this again.. I'm guessing that by having access to snapshots I'll be able to use that to materialise the data I need.. however it does seen a bit overkill. I'd prefer if rebus sage's had a way to soft delete sagas by marking them as complete instead of hard delete. Then I could later add a UI over the top of same tables and this saves replicating state around.
